### PR TITLE
Add support for MQTT 3.1.1

### PR DESCRIPTION
**Pull request by alexandruioanp**
This PR adds the changes needed for MQTT 3.1.1 (protocol level 4) support as described [here](https://github.com/mqtt/mqtt.github.io/wiki/Differences-between-3.1.0-and-3.1.1).

Based on my understanding of the UTF-8 requirements, the client code should be compliant as-is.

### CODE REVIEW

File: examples/mqtt-client/mqtt-client.c [hunk 1]
```diff
@@ -351,7 +351,17 @@ mqtt_event(struct mqtt_connection *m, mqtt_event_t event, void *data)
     break;
   }
   case MQTT_EVENT_SUBACK: {
+    #if MQTT_311
```
Comment (g-oikonomou) [inline examples/mqtt-client/mqtt-client.c]: Typically we don't indent cpp

File: examples/mqtt-client/mqtt-client.c [hunk 2]
```diff
@@ -351,7 +351,17 @@ mqtt_event(struct mqtt_connection *m, mqtt_event_t event, void *data)
...
+    mqtt_suback_event_t *suback_event = (mqtt_suback_event_t*) data;
```
Comment (g-oikonomou) [inline examples/mqtt-client/mqtt-client.c]: code style: space before the *, no space after the closing paren. Try uncrustify scripts in `tools/`

File: examples/platform-specific/cc26x0-cc13x0/cc26x0-web-demo/mqtt-client.c [hunk 3]
```diff
@@ -469,7 +469,17 @@ mqtt_event(struct mqtt_connection *m, mqtt_event_t event, void *data)
     break;
   }
   case MQTT_EVENT_SUBACK: {
-    DBG("APP - Application is subscribed to topic successfully\n");
+    #if MQTT_311
```
Comment (g-oikonomou) [inline examples/platform-specific/cc26x0-cc13x0/cc26x0-web-demo/mqtt-client.c]: Ditto

File: examples/platform-specific/nrf52dk/mqtt-demo/mqtt-demo.c [hunk 4]
```diff
@@ -305,7 +305,17 @@ mqtt_event(struct mqtt_connection *m, mqtt_event_t event, void *data)
     break;
   }
   case MQTT_EVENT_SUBACK: {
-    DBG("APP - Application is subscribed to topic successfully\n");
+    #if MQTT_311
```
Comment (g-oikonomou) [inline examples/platform-specific/nrf52dk/mqtt-demo/mqtt-demo.c]: .

File: os/net/app-layer/mqtt/mqtt.c [hunk 5]
```diff
@@ -114,9 +114,52 @@ typedef enum {
...
   MQTT_VHDR_CONN_REJECTED_BAD_USER_PASS,
   MQTT_VHDR_CONN_REJECTED_UNAUTHORIZED,
-} mqtt_vhdr_connack_fields_t;
+} mqtt_vhdr_connack_ret_code_t;
+
+#if MQTT_PROTOCOL_VERSION >= MQTT_PROTOCOL_VERSION_3_1_1
+typedef enum {
+  MQTT_VHDR_CONNACK_SESSION_PRESENT = 0x1
+} mqtt_vhdr_connack_flags_t;
+#endif
```
Comment (g-oikonomou) [inline os/net/app-layer/mqtt/mqtt.c]: I see no reason why not declare the enum unconditionally

File: os/net/app-layer/mqtt/mqtt.c [hunk 6]
```diff
@@ -659,6 +702,11 @@ PT_THREAD(publish_pt(struct pt *pt, struct mqtt_connection *conn))
     PT_EXIT(pt);
   }
 
+  // The DUP flag MUST be set to 0 for all QoS 0 messages
```
Comment (g-oikonomou) [inline os/net/app-layer/mqtt/mqtt.c]: Use C comments (/* */)

File: os/net/app-layer/mqtt/mqtt.c [hunk 7]
```diff
@@ -1295,9 +1372,11 @@ mqtt_register(struct mqtt_connection *conn, struct process *app_process,
               char *client_id, mqtt_event_callback_t event_callback,
               uint16_t max_segment_size)
 {
+  #if MQTT_31 || !MQTT_SRV_SUPPORTS_EMPTY_CLIENT_ID
```
Comment (g-oikonomou) [inline os/net/app-layer/mqtt/mqtt.c]: .

File: os/net/app-layer/mqtt/mqtt.c [hunk 8]
```diff
@@ -1339,7 +1418,11 @@ mqtt_connect(struct mqtt_connection *conn, char *host, uint16_t port,
   conn->server_port = port;
   conn->out_buffer_ptr = conn->out_buffer;
   conn->out_packet.qos_state = MQTT_QOS_STATE_NO_ACK;
-  conn->connect_vhdr_flags |= MQTT_VHDR_CLEAN_SESSION_FLAG;
+
+  // If the Client supplies a zero-byte ClientId, the Client MUST also set CleanSession to 1
```
Comment (g-oikonomou) [inline os/net/app-layer/mqtt/mqtt.c]: .

File: os/net/app-layer/mqtt/mqtt.h [hunk 9]
```diff
@@ -113,9 +131,23 @@
...
+#endif
+
 #define MQTT_TOPIC_MAX_LENGTH 128
+
+#if MQTT_PROTOCOL_VERSION >= MQTT_PROTOCOL_VERSION_3_1_1
+#ifdef MQTT_CONF_SUPPORTS_EMPTY_CLIENT_ID
+#define MQTT_SRV_SUPPORTS_EMPTY_CLIENT_ID MQTT_CONF_SUPPORTS_EMPTY_CLIENT_ID
+#else
+#define MQTT_SRV_SUPPORTS_EMPTY_CLIENT_ID 0
+#endif
```
Comment (g-oikonomou) [inline os/net/app-layer/mqtt/mqtt.h]: Conventionally we'd add a `/* MQTT_CONF_SUPPORTS_EMPTY_CLIENT_ID */` to show which conditional is closing here

File: os/net/app-layer/mqtt/mqtt.h [hunk 10]
```diff
@@ -321,6 +362,10 @@ struct mqtt_connection {
   uint8_t connect_vhdr_flags;
   uint8_t auto_reconnect;
 
+#if MQTT_PROTOCOL_VERSION >= MQTT_PROTOCOL_VERSION_3_1_1
+  uint8_t session_present;
+#endif
+
```
Comment (g-oikonomou) [inline os/net/app-layer/mqtt/mqtt.h]: Likely to cause struct padding (or is it?). Poss to avoid?

**Comment by alexandruioanp** [comment]
I have a commit that adds an execution test for V3.1.1 but it looks like it might have to wait for #1061.

**Comment by alexandruioanp** [comment]
#1062 must be merged first.

**Comment by g-oikonomou** [comment]
Valgrind tests are really cool and it looks like the test is passing nicely with #1062. Only thing that's left AFAIK is the check for NULL chars inside strings as discussed offline.

**Comment by g-oikonomou** [comment]
Anyone any idea why this is failing? I restarted but it failed in the same fashion again.

```
==== 06-native-coap.log ====
2315Get .well-known/core
2316Invalid URL. Protocol is not given or URL is malformed.
2317Get test/push
2318Invalid URL. Protocol is not given or URL is malformed.
231906-native-coap                   TEST FAIL    0/2
```

https://travis-ci.org/contiki-ng/contiki-ng/jobs/585770583

**Comment by alexandruioanp** [comment]
That's weird, the test passes fine in my local container.

**Pull request merged**